This Case Study is targeted to build a recommendation system
based on individual consumer’s behaviour or choice.
DATA DESCRIPTION:
• author : name of the person who gave the rating
• country : country the person who gave the rating belongs to
• data : date of the rating
• domain: website from which the rating was taken from
• extract: rating content
• language: language in which the rating was given
• product: name of the product/mobile phone for which the rating was given
• score: average rating for the phone
• score_max: highest rating given for the phone
• source: source from where the rating was taken

In [1]:
import os
import numpy as np  
import pandas as pd
#  To enable plotting graphs in Jupyter notebook
%matplotlib inline 
# Numerical libraries
 

# importing ploting libraries
import matplotlib.pyplot as plt   
import matplotlib.style
plt.style.use('classic')

#importing seaborn for statistical plots
import seaborn as sns

from scipy.stats import zscore

Reading give csv files

In [2]:
data1 = pd.read_csv("phone_user_review_file_1.csv",encoding='ISO-8859–1') 

In [3]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374910 entries, 0 to 374909
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   phone_url  374910 non-null  object 
 1   date       374910 non-null  object 
 2   lang       374910 non-null  object 
 3   country    374910 non-null  object 
 4   source     374910 non-null  object 
 5   domain     374910 non-null  object 
 6   score      366691 non-null  float64
 7   score_max  366691 non-null  float64
 8   extract    371934 non-null  object 
 9   author     371641 non-null  object 
 10  product    374910 non-null  object 
dtypes: float64(2), object(9)
memory usage: 31.5+ MB


In [4]:
data2 = pd.read_csv("phone_user_review_file_2.csv",encoding='ISO-8859–1') 
data3 = pd.read_csv("phone_user_review_file_3.csv",encoding='ISO-8859–1') 
data4 = pd.read_csv("phone_user_review_file_4.csv",encoding='ISO-8859–1') 
data5 = pd.read_csv("phone_user_review_file_5.csv",encoding='ISO-8859–1') 
data6 = pd.read_csv("phone_user_review_file_6.csv",encoding='ISO-8859–1') 

In [5]:
data1.head(10)

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   
5  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
6  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
7  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
8  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
9  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   
5  verizonwireless.com   10.0       10.0   
6  verizonwireless.com    4.0       10.0   
7  verizonwireless.com    6.0       10.0   
8  verizonwireless.com    6.0       10.0   
9  verizonwireless.com    8.0       10.0   

                                             extract           author  \
0  As a diehard Samsung fan who has had every Sam...       CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...        james0923   
2  Adequate feel. Nice heft. Processor's still sl...         R. Craig   
3  Never disappointed. One of the reasons I've be...      Buster2020    
4  I've now found that i'm in a group of people t...       S Ate Mine   
5  I am the type of person who never would comple...            BDB76   
6  The way this Samsung S8 phone operates is more...         KLC30306   
7  I bought this phone very excited to use it. I ...             Mnhy   
8  It is an extremely advanced and truly a Smart ...          Mr Alan   
9  Great phone with a PHENOMENAL Camera, not all ...  Raven Wolf Song   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8  
5                                  Samsung Galaxy S8  
6                                  Samsung Galaxy S8  
7                                  Samsung Galaxy S8  
8                                  Samsung Galaxy S8  
9                                  Samsung Galaxy S8

In [6]:
data2.head(10)

phone_url        date lang country  source        domain  \
0  /cellphones/leagoo-lead-7/   4/15/2015   en      us  Amazon    amazon.com   
1  /cellphones/leagoo-lead-7/   5/23/2015   en      gb  Amazon  amazon.co.uk   
2  /cellphones/leagoo-lead-7/   4/27/2015   en      gb  Amazon  amazon.co.uk   
3  /cellphones/leagoo-lead-7/   4/22/2015   en      gb  Amazon  amazon.co.uk   
4  /cellphones/leagoo-lead-7/   4/18/2015   en      gb  Amazon  amazon.co.uk   
5  /cellphones/leagoo-lead-7/   4/14/2015   en      us  Amazon    amazon.com   
6  /cellphones/leagoo-lead-7/    4/1/2015   en      gb  Amazon  amazon.co.uk   
7  /cellphones/leagoo-lead-7/   11/6/2015   it      it  Amazon     amazon.it   
8  /cellphones/leagoo-lead-7/  10/26/2015   it      it  Amazon     amazon.it   
9  /cellphones/leagoo-lead-7/   9/19/2015   it      it  Amazon     amazon.it   

   score  score_max                                            extract  \
0    2.0       10.0  The telephone headset is of poor quality , not...   
1   10.0       10.0  This is my first smartphone so I have nothing ...   
2    8.0       10.0  Great phone. Battery life not great but seems ...   
3   10.0       10.0      Best 90 quid I've ever spent on a smart phone   
4   10.0       10.0  I m happy with this phone.it s very good.thx team   
5   10.0       10.0  Exelent phone. Evriting perfect. Service is am...   
6   10.0       10.0  Great mobile, but contains malware, in the for...   
7   10.0       10.0  maneggevole, pratico, veloce, ottime foto........   
8   10.0       10.0  Prodotto ECCELLENTE!!! Praticamente perfetto, ...   
9   10.0       10.0  Cercavo un telefono che avesse una fotocamera ...   

          author                                            product  
0           luis  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1     Mark Lavin  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2         tracey  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3  Reuben Ingram  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4         viorel  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
5         Milena  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
6    lee carroll  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
7     Filippo L.  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...  
8    Emiliano S.  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...  
9        rossano  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...

In [7]:
data3.head(10)

phone_url       date lang country  \
0  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt      br   
1  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt      br   
2  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
3  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
4  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   
5  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   
6  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   8/3/2015   pt      br   
7  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/29/2015   pt      br   
8  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/24/2015   pt      br   
9  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/21/2015   pt      br   

      source            domain  score  score_max  \
0  Submarino  submarino.com.br    6.0       10.0   
1  Submarino  submarino.com.br   10.0       10.0   
2  Submarino  submarino.com.br   10.0       10.0   
3  Submarino  submarino.com.br    8.0       10.0   
4    Colombo    colombo.com.br    8.0       10.0   
5  Submarino  submarino.com.br    8.0       10.0   
6  Submarino  submarino.com.br   10.0       10.0   
7  Submarino  submarino.com.br    4.0       10.0   
8  Submarino  submarino.com.br   10.0       10.0   
9  Submarino  submarino.com.br   10.0       10.0   

                                             extract  \
0  recomendo, eu comprei um, a um ano, e agora co...   
1  Comprei um pouco desconfiada do site e do celu...   
2  Muito bom o produto, obvio que tem versões me...   
3  Unica ressalva fica para a camera que poderia ...   
4  Rapidez e atenção na entrega. O aparelho é ...   
5  Gostei do produto e entrega ocorreu antes do p...   
6  ótimo aparelho, unica coisa ruim é a camera ...   
7  Aparelho muito simples pelo preço que se paga...   
8  excelente produto gostei muito, estou satifeit...   
9  È um excelente aparelho,leve prático ....câ...   

                           author  \
0                herlington tesch   
1             Luisa Silva Marieta   
2                           Cyrus   
3       Marcela Santa Clara Brito   
4  Claudine Maria Kuhn Walendorff   
5                            JCVS   
6                   Juliana Viana   
7                     Julio cabal   
8         RIVALDO CIOQUETA JUNIOR   
9                         Dulcili   

                                             product  
0  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4  Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...  
5  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
6  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
7  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
8  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
9  Samsung Smartphone Samsung Galaxy S3 Slim G381...

In [8]:
data4.head(10)

phone_url        date lang country  \
0  /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1  /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
5  /cellphones/samsung-s7262-duos-galaxy-ace/  27/10/2015   en      in   
6  /cellphones/samsung-s7262-duos-galaxy-ace/  27/10/2015   en      us   
7  /cellphones/samsung-s7262-duos-galaxy-ace/  26/10/2015   en      in   
8  /cellphones/samsung-s7262-duos-galaxy-ace/  25/10/2015   en      in   
9  /cellphones/samsung-s7262-duos-galaxy-ace/  21/10/2015   en      in   

   source      domain  score  score_max  \
0  Amazon  amazon.com    2.0       10.0   
1  Zopper  zopper.com   10.0       10.0   
2  Amazon   amazon.in    4.0       10.0   
3  Amazon   amazon.in    6.0       10.0   
4  Amazon   amazon.in   10.0       10.0   
5  Amazon   amazon.in    8.0       10.0   
6  Amazon  amazon.com    2.0       10.0   
7  Amazon   amazon.in    2.0       10.0   
8  Amazon   amazon.in    6.0       10.0   
9  Amazon   amazon.in   10.0       10.0   

                                             extract            author  \
0  was not conpatable with my phone as stated. I ...  Frances DeSimone   
1  Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2  Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                               not bad for features   Amazon Customer   
4                                  Excellent product               NHK   
5                           Good in reasonable price            Rupali   
6  I hate it no flash no front camera microphone ...            Ashney   
7  The phone is getting struck very frequently.. ...   Amazon Customer   
8  Original product received with complete packag...   Amazon Customer   
9  Good product from Samsung. Just right for ever...   Amazon Customer   

                                             product  
0  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                Samsung Galaxy Star Pro S7262 Black  
2  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
5  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
6  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
7  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
8  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
9  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)

In [9]:
data5.head(10)

phone_url        date lang country      source  \
0  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/   4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/   4/23/2013   en      in     Naaptol   
5  /cellphones/karbonn-k1616/   2/24/2013   en      in     Naaptol   
6  /cellphones/karbonn-k1616/   2/23/2013   en      in     Naaptol   
7  /cellphones/karbonn-k1616/    2/7/2013   en      in     Naaptol   
8  /cellphones/karbonn-k1616/   1/12/2013   en      in     Naaptol   
9  /cellphones/karbonn-k1616/  10/13/2012   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   
5    naaptol.com    8.0       10.0   
6    naaptol.com   10.0       10.0   
7    naaptol.com    8.0       10.0   
8    naaptol.com    8.0       10.0   
9    naaptol.com    8.0       10.0   

                                             extract             author  \
0  I bought 1 month before. currently speaker is ...          venkatesh   
1  I just bought one week back, I have Airtel con...             Venkat   
2  one problem in this handset opera is not worki...             krrish   
3  here Karbonn comes up with an another excellen...    BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...      Suraj CHAUHAN   
5  here Karbonn comes up with an another excellen...      Prasoon Gupta   
6  the camera quality is good compared even with ...    Yogeet Patnayak   
7  this mobile having basic features with compare...       Murali Avula   
8  Karbonn mobiles are well known for there quali...  Jeseel Cholassery   
9  The phone is handy, good build quality, howeve...      Sobhit tiwari   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black  
5  Karbonn K1616 - Black  
6  Karbonn K1616 - Black  
7  Karbonn K1616 - Black  
8  Karbonn K1616 - Black  
9  Karbonn K1616 - Black

In [10]:
data6.head(10)

phone_url        date lang country  \
0  /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1  /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2  /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3  /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4  /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   
5  /cellphones/samsung-instinct-sph-m800/   1/28/2009   en      us   
6  /cellphones/samsung-instinct-sph-m800/   7/25/2008   en      us   
7  /cellphones/samsung-instinct-sph-m800/   5/12/2015   en      us   
8  /cellphones/samsung-instinct-sph-m800/   4/26/2015   en      us   
9  /cellphones/samsung-instinct-sph-m800/   9/11/2014   en      us   

                      source             domain  score  score_max  \
0                Phone Arena     phonearena.com    8.0       10.0   
1                     Amazon         amazon.com    6.0       10.0   
2                Phone Scoop     phonescoop.com    9.0       10.0   
3              HandCellPhone  handcellphone.com    4.0       10.0   
4               Reviewed.com       reviewed.com    6.0       10.0   
5               ReviewStream   reviewstream.com    NaN        NaN   
6  Personal Electronics buzz         pebuzz.com    NaN        NaN   
7                     Amazon         amazon.com    4.0       10.0   
8                     Amazon         amazon.com    6.0       10.0   
9                     Amazon         amazon.com    4.0       10.0   

                                             extract           author  \
0  I've had the phone for awhile and it's a prett...       ajabrams95   
1  to be clear it is not the sellers fault that t...        Stephanie   
2  Well i love this phone. i have had ton of phon...         snickers   
3  I have had my Instinct for several months now ...              A4C   
4  i have had this instinct phone for about two m...        betaBgood   
5  When I heard that my friend had gotten the new...             rich   
6  This is a great multi-media phone. The screen ...        Highlight   
7  For the price I expected the phone not to be s...          jessica   
8          Good for a temporary phone for a preteen.  Amazon Customer   
9          Didn't work. Helpful in getting a refund.  Tamarra Kemsley   

                     product  
0        Samsung Instinct HD  
1  Samsung SPH M800 Instinct  
2              Instinct M800  
3           Samsung Instinct  
4           Samsung Instinct  
5     Samsung Instinct Phone  
6           Samsung Instinct  
7  Samsung SPH M800 Instinct  
8  Samsung SPH M800 Instinct  
9  Samsung SPH M800 Instinct

Checking scores with 0 values in data

In [11]:
c=0
for m in data1['score']:
    if m==0:
        c=c+1
        print(c)
        
c=0
for m in data2['score']:
    if m==0:
        c=c+1
        print(c)
c=0
for m in data3['score']:
    if m==0:
        c=c+1
        print(c)
        
c=0
for m in data4['score']:
    if m==0:
        c=c+1
        print(c)        

c=0
for m in data5['score']:
    if m==0:
        c=c+1
        print(c)        
c=0
for m in data6['score']:
    if m==0:
        c=c+1
        print(c)        

Did not found any score with 0 value

Rounding off scores to the nearest integers

In [12]:
data1['score']=data1['score'].round()
data2['score']=data2['score'].round()
data3['score']=data3['score'].round()
data4['score']=data4['score'].round()
data5['score']=data5['score'].round()
data6['score']=data6['score'].round()

In [13]:
print(data2.info())
print(data3.info())
print(data4.info())
print(data5.info())
print(data6.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114925 entries, 0 to 114924
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   phone_url  114925 non-null  object 
 1   date       114925 non-null  object 
 2   lang       114925 non-null  object 
 3   country    114925 non-null  object 
 4   source     114925 non-null  object 
 5   domain     114925 non-null  object 
 6   score      112166 non-null  float64
 7   score_max  112166 non-null  float64
 8   extract    113965 non-null  object 
 9   author     113290 non-null  object 
 10  product    114925 non-null  object 
dtypes: float64(2), object(9)
memory usage: 9.6+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 312961 entries, 0 to 312960
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   phone_url  312961 non-null  object 
 1   date       312961 non-null  object 
 2   lang       312961 n

In [14]:
print(data1.isnull().sum())

phone_url       0
date            0
lang            0
country         0
source          0
domain          0
score        8219
score_max    8219
extract      2976
author       3269
product         0
dtype: int64


Found null values in data

Checking for values with score max greater than 10

In [15]:
print(data1[['score_max']].loc[data1['score_max']<10])
print(data1[['score_max']].loc[data1['score_max']>10])

Empty DataFrame
Columns: [score_max]
Index: []
Empty DataFrame
Columns: [score_max]
Index: []


Did not found any score max greater than and less than 10

Filling missing score with 0 and missing score max with

In [16]:
data1['score_max']=data1['score_max'].fillna(10)
data1['score']=data1['score'].fillna(0)
data1['author']=data1['author'].fillna('NA')
data1['extract']=data1['extract'].fillna('NA')

In [17]:
print(data1.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


In [18]:
data2['score_max']=data2['score_max'].fillna(10)
data2['score']=data2['score'].fillna(0)
data2['author']=data2['author'].fillna('NA')
data2['extract']=data2['extract'].fillna('NA')
print(data2.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


In [19]:
data3['score_max']=data3['score_max'].fillna(10)
data3['score']=data3['score'].fillna(0)
data3['author']=data3['author'].fillna('NA')
data3['extract']=data3['extract'].fillna('NA')
data3=data3.dropna(axis=0)
print(data3.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


In [20]:
data4['score_max']=data4['score_max'].fillna(10)
data4['score']=data4['score'].fillna(0)
data4['author']=data4['author'].fillna('NA')
data4['extract']=data4['extract'].fillna('NA')
print(data4.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


In [21]:
data5['score_max']=data5['score_max'].fillna(10)
data5['score']=data5['score'].fillna(0)
data5['author']=data5['author'].fillna('NA')
data5['extract']=data5['extract'].fillna('NA')
print(data5.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


In [22]:
data6['score_max']=data6['score_max'].fillna(10)
data6['score']=data6['score'].fillna(0)
data6['author']=data6['author'].fillna('NA')
data6['extract']=data6['extract'].fillna('NA')
print(data6.isnull().sum())

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64


Merging the provided CSVs into one data-frame

In [23]:
result=data1.append(data2)

In [24]:
result=result.append(data3)
result=result.append(data4)
result=result.append(data5)
result=result.append(data6)

In [25]:
result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1415132 entries, 0 to 163836
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1415132 non-null  object 
 1   date       1415132 non-null  object 
 2   lang       1415132 non-null  object 
 3   country    1415132 non-null  object 
 4   source     1415132 non-null  object 
 5   domain     1415132 non-null  object 
 6   score      1415132 non-null  float64
 7   score_max  1415132 non-null  float64
 8   extract    1415132 non-null  object 
 9   author     1415132 non-null  object 
 10  product    1415132 non-null  object 
dtypes: float64(2), object(9)
memory usage: 129.6+ MB


Resetting index in dataframe

In [26]:
result.reset_index(inplace=True)

In [27]:
result.drop(['index'], axis='columns', inplace=True)

In [28]:
result.drop_duplicates(inplace=True)

Dropping duplicate rows

In [29]:
duplicate= result.duplicated()

In [30]:
duplicate

0          False
1          False
2          False
3          False
4          False
           ...  
1415127    False
1415128    False
1415129    False
1415130    False
1415131    False
Length: 1408696, dtype: bool

In [31]:
duplicate.any()

False

In [32]:
for i in duplicate.index:
    if duplicate[i]:
        print(i)
        print(result.iloc[i])

In [33]:
sampleData=result.sample(n=1000000, random_state=612)

In [34]:
sampleData.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000000 entries, 430175 to 242141
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1000000 non-null  object 
 1   date       1000000 non-null  object 
 2   lang       1000000 non-null  object 
 3   country    1000000 non-null  object 
 4   source     1000000 non-null  object 
 5   domain     1000000 non-null  object 
 6   score      1000000 non-null  float64
 7   score_max  1000000 non-null  float64
 8   extract    1000000 non-null  object 
 9   author     1000000 non-null  object 
 10  product    1000000 non-null  object 
dtypes: float64(2), object(9)
memory usage: 91.6+ MB


Dropping unwanted features and keeping only 1000000 data samples. Using random state=612.

In [35]:
final_data=sampleData.drop(['phone_url','date','lang','country','source','domain','score_max','extract'], axis='columns')

In [36]:
final_data

score                  author  \
430175     8.0  Vinod Kumar Chengespur   
587153     2.0                  Sharon   
594786     8.0             An Sionnach   
636378     2.0             Jomine Jose   
684531    10.0           Marcos Rivero   
...        ...                     ...   
1002309    6.0            Андрей   
1111094    8.0           GEORGESDEBOTE   
243653    10.0                   e-bit   
92037      8.0         Client d'Amazon   
242141     4.0                    JYLA   

                                                   product  
430175                    Lenovo Vibe K4 Note (White,16GB)  
587153   HTC Desire 816 Black (Virgin mobile) - 5.5 inc...  
594786   Xperia Sony Xperia Z Ultra SIM-free Android Sm...  
636378                                Lava Iris X5 (Black)  
684531   Sony Xperia L C2105 White (Factory Unlocked) 4...  
...                                                    ...  
1002309                           Samsung S5830 Galaxy Ace  
1111094                                  Sony Ericsson txt  
243653   Smartphone Motorola Moto X 2ª Geração XT109...  
92037    Buyus X2 Film Protecteur d'écran VITRE en VER...  
242141       Samsung Smartphone Samsung Galaxy J7 16 Go Or  

[1000000 rows x 3 columns]

# 2. Most rated 5 products are given below

In [37]:
final_data['product'].value_counts().sort_values(ascending=False).head()  

Lenovo Vibe K4 Note (White,16GB)     3709
Lenovo Vibe K4 Note (Black, 16GB)    3083
OnePlus 3 (Graphite, 64 GB)          2890
OnePlus 3 (Soft Gold, 64 GB)         2522
Huawei P8lite zwart / 16 GB          1895
Name: product, dtype: int64

# The users with most number of reviews

In [38]:
final_data['author'].value_counts().sort_values(ascending=False).head() 

Amazon Customer    54641
NA                 43867
Cliente Amazon     13669
e-bit               5962
Client d'Amazon     5499
Name: author, dtype: int64

In [39]:
prodStore=final_data['product'].value_counts()
prodStore

Lenovo Vibe K4 Note (White,16GB)                                                                                           3709
Lenovo Vibe K4 Note (Black, 16GB)                                                                                          3083
OnePlus 3 (Graphite, 64 GB)                                                                                                2890
OnePlus 3 (Soft Gold, 64 GB)                                                                                               2522
Huawei P8lite zwart / 16 GB                                                                                                1895
                                                                                                                           ... 
THANSITA New Samsung Galaxy Note 3 Sm-n9005 Quad 5.7'' 13mp 4g LTE 32gb Black/gold Phone Fast Shipping                        1
Samsung Galaxy S7 Edge Smartphone débloqué 4G (Ecran : 5,5 pouces - 32 Go - 4 Go RAM - Android) Blanc 

In [40]:
authStore=final_data['author'].value_counts()
authStore

Amazon Customer      54641
NA                   43867
Cliente Amazon       13669
e-bit                 5962
Client d'Amazon       5499
                     ...  
R.H. "Pepsionice"        1
Backlash                 1
marcos fernandes         1
Raptorhn                 1
victoria price           1
Name: author, Length: 596751, dtype: int64

In [41]:
prod_50=prodStore.loc[prodStore>50]

In [42]:
auth_50=authStore.loc[authStore>50]

In [45]:
prod_50

Lenovo Vibe K4 Note (White,16GB)                                                                                   3709
Lenovo Vibe K4 Note (Black, 16GB)                                                                                  3083
OnePlus 3 (Graphite, 64 GB)                                                                                        2890
OnePlus 3 (Soft Gold, 64 GB)                                                                                       2522
Huawei P8lite zwart / 16 GB                                                                                        1895
                                                                                                                   ... 
Samsung Nexus S GT-I9023                                                                                             51
Sony Xperia XCompact Smartphone (11,7 cm (4,6 Zoll), 32 GB Speicher, Android 6.0) Mist Blue                          51
Huawei P8 Lite - Smartphone libre Androi

In [49]:
prod_grp=final_data.loc[final_data['product']==prod_50.index[0]]
prod_grp=prod_grp.append(final_data.loc[final_data['product']==prod_50.index[1]])
prod_grp.drop_duplicates(inplace=True)
prod_grp

score                   author                            product
430175    8.0   Vinod Kumar Chengespur   Lenovo Vibe K4 Note (White,16GB)
428727   10.0  Chiranjivi Girish Thuse   Lenovo Vibe K4 Note (White,16GB)
423897    8.0                    sunil   Lenovo Vibe K4 Note (White,16GB)
428677    2.0            Animesh Ghosh   Lenovo Vibe K4 Note (White,16GB)
432164    2.0          Amazon Customer   Lenovo Vibe K4 Note (White,16GB)
...       ...                      ...                                ...
424299    2.0                 jkchavda  Lenovo Vibe K4 Note (Black, 16GB)
426903   10.0         rajesh devisetti  Lenovo Vibe K4 Note (Black, 16GB)
429962   10.0          Rushikesh Kolhe  Lenovo Vibe K4 Note (Black, 16GB)
426447   10.0               Rietesh P.  Lenovo Vibe K4 Note (Black, 16GB)
431561    2.0  Racharla Vamshi Krishna  Lenovo Vibe K4 Note (Black, 16GB)

[2848 rows x 3 columns]

In [56]:
prod_grp.shape

(155135, 3)

In [54]:
prod_grp.drop_duplicates(inplace=True)

In [55]:
prod_grp

score                   author                           product
430175     8.0   Vinod Kumar Chengespur  Lenovo Vibe K4 Note (White,16GB)
428727    10.0  Chiranjivi Girish Thuse  Lenovo Vibe K4 Note (White,16GB)
423897     8.0                    sunil  Lenovo Vibe K4 Note (White,16GB)
428677     2.0            Animesh Ghosh  Lenovo Vibe K4 Note (White,16GB)
432164     2.0          Amazon Customer  Lenovo Vibe K4 Note (White,16GB)
...        ...                      ...                               ...
1091798   10.0             Naveen Power                 Sony Ericsson Elm
1092224   10.0        giblartar.vasilij                 Sony Ericsson Elm
1092010   10.0              dima1dima-9                 Sony Ericsson Elm
1092120    8.0              BOULANGERAC                 Sony Ericsson Elm
1092161   10.0          GeorgeSavchenko                 Sony Ericsson Elm

[155135 rows x 3 columns]

In [57]:
prod_grp=final_data.loc[final_data['product']==prod_50.index[0]]
for p in range(1,4256):
    prod_grp=prod_grp.append(final_data.loc[final_data['product']==prod_50.index[p]])
    prod_grp.drop_duplicates(inplace=True)

In [58]:
for p in auth_50.index:
    prod_grp=prod_grp.append(final_data.loc[final_data['author']==p])
    prod_grp.drop_duplicates(inplace=True)

In [59]:
prod_grp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 524655 entries, 430175 to 653242
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   score    524655 non-null  float64
 1   author   524655 non-null  object 
 2   product  524655 non-null  object 
dtypes: float64(1), object(2)
memory usage: 16.0+ MB


In [61]:
prod_grp.drop_duplicates(inplace=True)

# The data with products having more than 50 ratings and users who have given more than 50 ratings

In [62]:
prod_grp

score                   author  \
430175     8.0   Vinod Kumar Chengespur   
428727    10.0  Chiranjivi Girish Thuse   
423897     8.0                    sunil   
428677     2.0            Animesh Ghosh   
432164     2.0          Amazon Customer   
...        ...                      ...   
1363771    9.0                   Jordan   
1384457    6.0                   Jordan   
744205    10.0                   Jordan   
1363333    9.0                   Jordan   
653242     8.0                   Jordan   

                                          product  
430175           Lenovo Vibe K4 Note (White,16GB)  
428727           Lenovo Vibe K4 Note (White,16GB)  
423897           Lenovo Vibe K4 Note (White,16GB)  
428677           Lenovo Vibe K4 Note (White,16GB)  
432164           Lenovo Vibe K4 Note (White,16GB)  
...                                           ...  
1363771                          Samsung SCH-A930  
1384457                 Motorola V325i Cell Phone  
744205   HTC One Mini SIM-Free Smartphone - Black  
1363333                             Siemens SXG75  
653242           HTC One M7, Silver 32GB (Sprint)  

[524655 rows x 3 columns]

# The shape of the final dataset is given is 524655 rows × 3 columns

In [63]:
prod_grp.groupby('product')['score'].mean().sort_values(ascending=False).head()  

product
Nokia 1200 blue (Organizer, 3 Spiele) Handy                                                        10.0
Nokia Lumia 1520 32GB 4G Nero                                                                      10.0
Nokia Lumia 635 Smartphone, 8 GB, Bianco [Italia]                                                  10.0
Nokia Lumia 635 8GB Unlocked GSM 4G LTE Windows 8.1 Quad-Core Phone - White                        10.0
Nokia Lumia 635 8GB Unlocked GSM 4G LTE Windows 8.1 Quad-Core Phone - Nero,cassa del caso cover    10.0
Name: score, dtype: float64

# The top 5 mobile phones from popularity based model

Nokia 1200 blue (Organizer, 3 Spiele) Handy                                                       
Nokia Lumia 1520 32GB 4G Nero                                                                     
Nokia Lumia 635 Smartphone, 8 GB, Bianco [Italia]                                                  
Nokia Lumia 635 8GB Unlocked GSM 4G LTE Windows 8.1 Quad-Core Phone - White                        
Nokia Lumia 635 8GB Unlocked GSM 4G LTE Windows 8.1 Quad-Core Phone - Nero,cassa del caso cover 

In [64]:
from sklearn.preprocessing import LabelEncoder   # import label encoder

labelUser = LabelEncoder()
labelProd = LabelEncoder()
prod_grp['UserId'] = labelUser.fit_transform(prod_grp.author)


Building a collaborative filtering model using SVD

Labeling the authors and products

In [66]:
prod_grp['ProductId'] = labelProd.fit_transform(np.ravel(prod_grp['product']))
      
# np.ravel(prod_grp['product']).shape

In [85]:
prod_grp1=prod_grp.drop(['author','product'], axis='columns')

In [86]:
prod_grp1

score  UserId  ProductId
430175     8.0  213820      11224
428727    10.0   39451      11224
423897     8.0  317310      11224
428677     2.0   16857      11224
432164     2.0   13111      11224
...        ...     ...        ...
1363771    9.0  101324      22045
1384457    6.0  101324      13342
744205    10.0  101324       6668
1363333    9.0  101324      23080
653242     8.0  101324       6531

[524655 rows x 3 columns]

In [68]:
from collections import defaultdict

from surprise import SVD
from surprise import Dataset
from surprise import Reader

In [69]:
reader = Reader(rating_scale=(0, 10))

In [70]:
data = Dataset.load_from_df(prod_grp1[['UserId', 'ProductId', 'score']], reader)

In [71]:
from surprise.model_selection import train_test_split

trainset, testset = train_test_split(data, test_size=.2)

In [72]:
algo = SVD()

In [73]:
algo.fit(trainset)
predictions = algo.test(testset)

In [74]:
predictions

[Prediction(uid=332108, iid=9813, r_ui=8.0, est=7.619619032180249, details={'was_impossible': False}),
 Prediction(uid=49526, iid=2515, r_ui=2.0, est=5.796387461338838, details={'was_impossible': False}),
 Prediction(uid=203394, iid=1949, r_ui=10.0, est=9.40155369888648, details={'was_impossible': False}),
 Prediction(uid=36246, iid=13948, r_ui=10.0, est=8.69453294762487, details={'was_impossible': False}),
 Prediction(uid=89720, iid=20077, r_ui=8.0, est=8.199360247576516, details={'was_impossible': False}),
 Prediction(uid=320884, iid=10338, r_ui=4.0, est=6.954473492861082, details={'was_impossible': False}),
 Prediction(uid=139693, iid=17598, r_ui=8.0, est=8.697433482198994, details={'was_impossible': False}),
 Prediction(uid=135348, iid=12912, r_ui=10.0, est=8.026006304929567, details={'was_impossible': False}),
 Prediction(uid=64703, iid=3469, r_ui=4.0, est=7.091815000346537, details={'was_impossible': False}),
 Prediction(uid=68522, iid=14232, r_ui=2.0, est=8.366940364775635, deta

In [75]:
from surprise import accuracy
accuracy.rmse(predictions)

RMSE: 2.6713


2.6712704246754067

In [119]:
# trainset.ur

Build a collaborative filtering model using kNNWithMeans from surprise

Due to memory constraints 10000 samples are only taken

In [111]:
prod_grp2=prod_grp1.sample(n=10000, random_state=612)
data = Dataset.load_from_df(prod_grp2[['UserId', 'ProductId', 'score']], reader)
trainset, testset = train_test_split(data, test_size=.2)

In [112]:
from surprise import KNNWithMeans

In [113]:
algo1 = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline', 'user_based': True})
algo1.fit(trainset)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


In [118]:
# testset

In [115]:
test_pred = algo1.test(testset)

In [116]:
test_pred

[Prediction(uid=264150, iid=16556, r_ui=0.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=323441, iid=6977, r_ui=4.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=80427, iid=23973, r_ui=6.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=113029, iid=6256, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=243580, iid=14026, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=301675, iid=16595, r_ui=9.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=342186, iid=13897, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=60954, iid=15874, r_ui=4.0, est=7.732, details={'was_imp

In [117]:
# get RMSE
print("User-based Model : Test Set")
accuracy.rmse(test_pred, verbose=True)

User-based Model : Test Set
RMSE: 2.8789


2.8789349162009343

As less samples are taken for KNN model gives was impossible true for many entries in test set 

In [121]:
algo1 = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline', 'user_based': False})
algo1.fit(trainset)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


In [122]:
test_pred = algo1.test(testset)

In [123]:
test_pred

[Prediction(uid=264150, iid=16556, r_ui=0.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=323441, iid=6977, r_ui=4.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=80427, iid=23973, r_ui=6.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=113029, iid=6256, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=243580, iid=14026, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=301675, iid=16595, r_ui=9.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=342186, iid=13897, r_ui=8.0, est=7.732, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=60954, iid=15874, r_ui=4.0, est=7.732, details={'was_imp

In [124]:
# get RMSE
print("User-based Model : Test Set")
accuracy.rmse(test_pred, verbose=True)

User-based Model : Test Set
RMSE: 2.8569


2.8568871665597917

# Answer 10
Popularity based Recommendation Systems is used when no information is available about user preferences or user.
It is also used when user features are not related to any other user in training data set in case of user based recommendation.
In case of product based recommendation when product feature are not related to other products in the training data set also popularity based recommendation is used.

# Answer 11
CF based Recommendation Systems is used when relation found between about users or products.

# Answer 12
Content based Recommendation Systems can also be used for further improving the recommendation system.